selfdrive/ui/translations/auto_translate.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import sys
+import xml.etree.ElementTree as ET
+from typing import TextIO, cast
+
+import requests
+
+OPENAI_MODEL = "gpt-4"
+OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
+OPENAI_PROMPT = "You are a professional translator from English to {language} (ISO 639 language code)." + \
+                "The following sentence or word is in the GUI of a software called openpilot, translate it accordingly."
+
+
+def print_log(text: str) -> None:
+  print(text, file=sys.stderr)
+
+
+def translate_phrase(text: str, language: str) -> str:
+  response = requests.post(
+    "https://api.openai.com/v1/chat/completions",
+    json={
+      "model": OPENAI_MODEL,
+      "messages": [
+        {
+          "role": "system",
+          "content": OPENAI_PROMPT.format(language=language),
+        },
+        {
+          "role": "user",
+          "content": text,
+        },
+      ],
+      "temperature": 0.8,
+      "max_tokens": 128,
+      "top_p": 1,
+    },
+    headers={
+      "Authorization": f"Bearer {OPENAI_API_KEY}",
+    },
+  )
+
+  response.raise_for_status()
+
+  data = response.json()
+
+  return cast(str, data["choices"][0]["message"]["content"])
+
+
+def translate_file(input_file: TextIO, output_file: TextIO, language: str, all_: bool) -> None:
+  tree = ET.parse(input_file)
+
+  root = tree.getroot()
+
+  root.attrib["language"] = language
+
+  for context in root.findall("./context"):
+    name = context.find("name")
+    if name is None:
+      raise ValueError("name not found")
+
+    print_log(f"Context: {name.text}")
+
+    for message in context.findall("./message"):
+      source = message.find("source")
+      translation = message.find("translation")
+
+      if source is None or translation is None:
+        raise ValueError("source or translation not found")
+
+      if not all_ and translation.attrib.get("type") != "unfinished":
+        continue
+
+      llm_translation = translate_phrase(cast(str, source.text), language)
+
+      print_log(f"Source: {source.text}\n" + \
+                f"Current translation: {translation.text}\n" + \
+                f"LLM translation: {llm_translation}")
+
+      translation.text = llm_translation
+
+  output_file.write('<?xml version="1.0" encoding="utf-8"?>\n' +
+                    '<!DOCTYPE TS>\n' +
+                    ET.tostring(root, encoding="utf-8").decode())
+
+
+def main() -> None:
+  arg_parser = argparse.ArgumentParser()
+  arg_parser.add_argument("input", type=argparse.FileType("r"), help="The input file")
```

Comment:
All of our other translation tools and tests assume the location of the files, can we do that here? This feels redundant for example:

`./auto_translate.py main_fr.ts -l fr`

Also, unclear what --all does, since it still requires the `input` and crashes if given everything:

```python
</TS>batman@workstation-shane:~/openpilot/selfdrive/ui/translations$ ./auto_translate.py main_fr.ts --all -l fr
Translates to fr (all)
Context: AbstractAlert
Traceback (most recent call last):
  File "/home/batman/openpilot/selfdrive/ui/translations/./auto_translate.py", line 109, in <module>
    main()
  File "/home/batman/openpilot/selfdrive/ui/translations/./auto_translate.py", line 105, in main
    translate_file(args.input, args.output, args.language, args.all)
  File "/home/batman/openpilot/selfdrive/ui/translations/./auto_translate.py", line 76, in translate_file
    llm_translation = translate_phrase(cast(str, source.text), language)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/openpilot/selfdrive/ui/translations/./auto_translate.py", line 45, in translate_phrase
    response.raise_for_status()
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.openai.com/v1/chat/completions
```

Reply:
Looks like a temporary problem with the API and not the code